### PR TITLE
TIAF Library Converson

### PR DESCRIPTION
**Pull request by jcurtisk**
## What does this PR do?

This PR refactors our main TIAF TestImpact script into a class structure, with specialized TestImpact objects for each runtime we are going to interact with. It also refactors persistent_storage and storage_query_tool into their own packages for cleanliness,

## How was this PR tested?

Suite of automated tests has been used to validate functionality has been retained and enhanced.

### CODE REVIEW

File: scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py [hunk 1]
```diff
@@ -395,6 +390,22 @@ def _generate_result(self, s3_bucket: str, suite: str, return_code: int, report:
...
+
+        @param args: Dictionary containing the arguments passed to this TestImpact object
+
+        @return: Compiled s3_top_level_dir name
+        """
+        if self.runtime_type:
+            if dir_name:
+                dir_name = os.path.join(dir_name, self.runtime_type)
+            else:
+                dir_name = os.path.join("tiaf", self.runtime_type)
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py:406]: I think we can remove the default `tiaf` and simply raise an exception if `s3_top_level_dir` is not specified when `s3` storage is used.

File: scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py [hunk 2]
```diff
@@ -395,6 +390,22 @@ def _generate_result(self, s3_bucket: str, suite: str, return_code: int, report:
         result["change_list"] = self._change_list
         return result
 
+    def _compile_s3_top_level_dir_name(self, dir_name: str):
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py:393]: We can remove `dir_name` as an arg from here.

File: scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py [hunk 3]
```diff
@@ -395,6 +390,22 @@ def _generate_result(self, s3_bucket: str, suite: str, return_code: int, report:
...
+        """
+        Function to build our s3_top_level_dir name. Reads the argument from our dictionary and then appends runtime_type to the end.
+        If s3_top_level_dir name is not provided in args, we will default to "tiaf"+runtime_type.
+
+        @param args: Dictionary containing the arguments passed to this TestImpact object
+
+        @return: Compiled s3_top_level_dir name
+        """
+        if self.runtime_type:
+            if dir_name:
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py:403]: We should use the `runtime_type` as the subdir beneath `s3_top_level_dir`.

File: scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py [hunk 4]
```diff
@@ -123,20 +125,13 @@ def _parse_arguments_to_runtime(self, args, sequence_type, runtime_args):
         runtime_args.append(f"--sequence={sequence_type}")
         logger.info(f"Sequence type is set to '{sequence_type}'.")
 
-        if args.get('safe_mode'):
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py:126]: Let's replace all string lits for args as named consts at the top of the file or class so we have an easy reference (also handy if they ever need to change).

**Review by jonawals — COMMENTED** [review]
Comments added as part of voice chat 👍

File: scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py (hunk 2 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/test_impact/base_test_impact.py:405]: Did not execute this change as I wanted to minimize passing the args dictionary around as much as possible,

File: scripts/build/TestImpactAnalysis/Testing/test_tiaf_unit_tests.py [hunk 5]
```diff
@@ -19,6 +19,26 @@ def assert_list_content_equal(list1, list2):
     assert Counter(list1) == Counter(list2)
 
 
+class ConcreteBaseTestImpact(BaseTestImpact):
+    """
+    Concrete implementation of BaseTestImpact so that we can execute unittests on the functionality of BaseTestImpact individually.
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/Testing/test_tiaf_unit_tests.py:24]: uber nit: unittests -> unit tests

**Review by hultonha — APPROVED** [review]
Updates look solid to me! Nice work @jcurtisk

File: scripts/build/TestImpactAnalysis/test_impact/test_impact_args.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+#
+
+from enum import Enum
+
+value = 0
+
+
+class RuntimeArgs(Enum):
```
[the remark this answers is not in the document]
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/test_impact/test_impact_args.py]: This is not possible in Python Enums unfortunately

File: scripts/build/TestImpactAnalysis/tiaf_driver.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
 import pathlib
```
[the remark this answers is not in the document]
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/tiaf_driver.py:12]: Will carry out in a separate PR.

**Pull request merged**